packages/block-library/src/table-of-contents/edit.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * Internal dependencies
+ */
+import * as Utils from './utils';
+import ListLevel from './ListLevel';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { subscribe } from '@wordpress/data';
+import {
+	IconButton,
+	Toolbar,
+	PanelBody,
+	ToggleControl,
+} from '@wordpress/components';
+import {
+	BlockControls,
+	InspectorControls,
+} from '@wordpress/editor';
+
+class TOCEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			wpDataUnsubscribe: null,
+		};
+
+		this.toggleAttribute = this.toggleAttribute.bind( this );
+		this.refresh = this.refresh.bind( this );
+	}
+
+	toggleAttribute( propName ) {
+		const value = this.props.attributes[ propName ];
+		const { setAttributes } = this.props;
+
+		setAttributes( { [ propName ]: ! value } );
+	}
+
+	refresh() {
+		const { setAttributes } = this.props;
+		const headings = Utils.getPageHeadings();
+		setAttributes( { headings } );
+	}
+
+	componentDidMount() {
+		const { attributes, setAttributes } = this.props;
+		const headings = attributes.headings || [];
+		const wpDataUnsubscribe = subscribe( () => {
```

Comment:
I'd suggest using `withSelect` to get the page headings instead of doing a manual subscribe here.

Reply:
@swissspidy Thanks for the suggestion. I will definitely look into that.